internal/worker/sshsession/worker.go

```diff
@@ -0,0 +1,363 @@
+	}
+	defer sshdConn.Close()
+
+	cancellableControllerConnection := newCancellableReadWriter(ctx, controllerConn)
+	cancellableSSHDConnection := newCancellableReadWriter(ctx, sshdConn)
+	eg := errgroup.Group{}
+
+	eg.Go(func() error {
+		// sshd -> conn
+		_, err := io.Copy(cancellableControllerConnection, cancellableSSHDConnection)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+
+	eg.Go(func() error {
+		// conn -> sshd
+		_, err = io.Copy(cancellableSSHDConnection, cancellableControllerConnection)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+
+	if err := eg.Wait(); err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// ConnectionGetter provides the methods to connect to a controller and the local SSHD.
+type ConnectionGetter interface {
+	GetControllerConnection(password, ctrlAddress string) (ssh.Channel, error)
+	GetSSHDConnection() (net.Conn, error)
+}
+
+// connectionGetter is capable of initating SSH connections to two places.
+//
+//  1. The controller's SSH server
+//  2. The local SSHD on the machine
+//
+// The consumer is expect to pipe these connections together.
+type connectionGetter struct {
+	logger Logger
+}
+
+// NewConnectionGetter returns a new connectionGetter.
+func NewConnectionGetter(l Logger) *connectionGetter {
+	return &connectionGetter{l}
+}
+
+// GetControllerConnection initiates an SSH connection to the target ctrlAddress.
+func (w *connectionGetter) GetControllerConnection(password, ctrlAddress string) (ssh.Channel, error) {
+	// TODO(ale8k): Watch will return host key in subsequent PR.
+	sshConfig := &ssh.ClientConfig{
+		User:            ControllerSSHUser,
+		Auth:            []ssh.AuthMethod{ssh.Password(password)},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(), // TODO(ale8k): Fill in host key here
+	}
+
+	client, err := ssh.Dial("tcp", ctrlAddress, sshConfig)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// TODO(ale8k): Make this a constant that both the server and session worker can use.
+	ch, in, err := client.OpenChannel("juju-tunnel", nil)
+	if err != nil {
+		return nil, err
+	}
+	go ssh.DiscardRequests(in)
+
+	return ch, nil
+}
+
+// GetSSHConnection performs a stand TCP dial to the SSHD running on the machine.
+func (w *connectionGetter) GetSSHDConnection() (net.Conn, error) {
+	etcPath := "/etc/ssh/sshd_config"
+	openSSHPath := "/usr/share/openssh/sshd_config"
+	port := w.getLocalSSHPort(etcPath, openSSHPath)
+	u, err := url.Parse("localhost" + port)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	localSSHD, err := net.Dial("tcp", u.String())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return localSSHD, nil
+}
+
+// getLocalSSHPort parses the local sshd_config file to get the port sshd is listening on.
+// It will try all the provided paths.
+// If it cannot get it for any reason, it'll log the error and return the default port 22.
+func (w *connectionGetter) getLocalSSHPort(filePaths ...string) string {
```

Comment:
I'm still not a big fan of doing this on every ssh connection. It feels like an extra hurdle in the overall flow when its unlikely the machine's SSH port will be changed frequently. But I'm okay to leave it like this.